### PR TITLE
Feature/python nuts params

### PR DESCRIPTION
**Pull request by christophernorth**
Made NUTS parameters settable from the python API (and slightly changed how the python "sample" API works so that it now takes an optional PosteriorSamplingAlgorithm object that represents the constructed sampler.

At the moment if the user doesn't specify an explicit sampler we just use metropolis hastings with default params, but this should probably change to use something a little more automagic (ie based on  differentiability).

Note - I haven't added customisable options for Hamiltonian just because it may go away.

### CODE REVIEW

File: keanu-python/tests/test_sampling.py [hunk 1]
```diff
@@ -161,28 +155,17 @@ def test_can_track_acceptance_rate_when_iterating(net: BayesNet) -> None:
...
-    with pytest.raises(TypeError) as excinfo:
-        sample(
-            net=net,
-            sample_from=net.get_latent_vertices(),
-            algo="hamiltonian",
-            proposal_listeners=[AcceptanceRateTracker()],
-            drop=3)
-    assert str(excinfo.value) == "Only Metropolis Hastings supports the proposal_listeners parameter"
 
+def test_can_specify_nuts_params(net: BayesNet):
```
Comment (wesleyw72) [inline keanu-python/tests/test_sampling.py]: nit: Can you add `-> None`

Comment (christophernorth) [reply keanu-python/tests/test_sampling.py]: 👍

File: keanu-python/keanu/algorithm/sampling.py [hunk 2]
```diff
@@ -17,31 +16,90 @@
...
+        self._sampler = sampler
+
+    def get_sampler(self) -> JavaObject:
+        return self._sampler
+
+
+class MetropolisHastingsSampler(PosteriorSamplingAlgorithm):
+
+    def __init__(self,
+                 proposal_distribution: str = None,
```
Comment (GeorgeNash) [inline keanu-python/keanu/algorithm/sampling.py]: I know that the `.builder()` sets all of the defaults defined in Java but they feel very hidden from a python user. What're your thoughts on adding all the defaults to the arguments here too? Means we have another thing to monitor across both API's but it'll be clearer for python users

File: keanu-python/keanu/algorithm/sampling.py [hunk 3]
```diff
@@ -17,31 +16,90 @@
...
+
+class HamiltonianSampler(PosteriorSamplingAlgorithm):
+
+    def __init__(self) -> None:
+        super().__init__(k.jvm_view().Hamiltonian.withDefaultConfig())
+
+
+class NUTSSampler(PosteriorSamplingAlgorithm):
+
+    def __init__(self,
```
Comment (GeorgeNash) [inline keanu-python/keanu/algorithm/sampling.py]: same comment about defaults

File: keanu-python/keanu/algorithm/sampling.py [hunk 4]
```diff
@@ -17,31 +16,90 @@
...
            draws: int = 500,
            drop: int = 0,
            down_sample_interval: int = 1,
            plot: bool = False,
            ax: Any = None) -> sample_types:
 
-    sampling_algorithm: JavaObject = build_sampling_algorithm(algo, proposal_distribution, proposal_distribution_sigma,
-                                                              proposal_listeners)
+    if sampling_algorithm is None:
+        sampling_algorithm = MetropolisHastingsSampler()
```
Comment (GeorgeNash) [inline keanu-python/keanu/algorithm/sampling.py]: add this to the `sample` method?

Comment (christophernorth) [reply keanu-python/keanu/algorithm/sampling.py]: Only if we want to change the semantics and share the "default" MetropolisHastingsSampler across all calls (which confuses some of the tests).  Also, it's generally considered bad form to create "mutable" objects as part of a default parameter in Python so I'd prefer to keep as is.

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/PosteriorSamplingAlgorithm.java [hunk 5]
```diff
@@ -22,4 +23,7 @@ default NetworkSamples getPosteriorSamples(BayesianNetwork bayesianNetwork, int
         return getPosteriorSamples(bayesianNetwork, bayesianNetwork.getTopLevelLatentVertices(), sampleCount);
     }
 
+    NetworkSamplesGenerator generatePosteriorSamples(final BayesianNetwork bayesianNetwork,
+                                                     final List<? extends Vertex> verticesToSampleFrom);
+
```
Comment (migwellian) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/PosteriorSamplingAlgorithm.java]: now you can give the other method a default version:

    default NetworkSamples getPosteriorSamples(BayesianNetwork bayesNet,
                                       List<? extends Vertex> verticesToSampleFrom,
                                       int sampleCount) {
        return generatePosteriorSamples(bayesNet, verticesToSampleFrom)
            .generate(sampleCount);
    }

Comment (christophernorth) [reply keanu-project/src/main/java/io/improbable/keanu/algorithms/PosteriorSamplingAlgorithm.java]: Good catch - will update

File: keanu-python/keanu/algorithm/sampling.py (hunk 2 above)
Comment (christophernorth) [reply keanu-python/keanu/algorithm/sampling.py]: As discussed - we should raise a ticket to document these things in Python better.  Deferring that till then.

File: keanu-python/keanu/algorithm/sampling.py (hunk 3 above)
Comment (christophernorth) [reply keanu-python/keanu/algorithm/sampling.py]: As above.

File: keanu-python/keanu/algorithm/sampling.py [hunk 6]
```diff
@@ -17,31 +16,90 @@
...
+                type_=proposal_distribution, sigma=proposal_distribution_sigma, listeners=proposal_listeners)
+            builder = builder.proposalDistribution(proposal_distribution_object.unwrap())
+
+        if use_cache_on_rejection is not None:
+            builder.useCacheOnRejection(use_cache_on_rejection)
+
+        super().__init__(builder.build())
+
+
+class HamiltonianSampler(PosteriorSamplingAlgorithm):
```
Comment (gordoncaleb) [inline keanu-python/keanu/algorithm/sampling.py]: We can just remove Hamilitonian. Might want to do it in another PR if it causes much extra effort.

Comment (christophernorth) [reply keanu-python/keanu/algorithm/sampling.py]: Yeah - I think it's best as its own PR.  Thanks for the review.
